File: evals/behavioral-analysis/data/csharp/injection-attacks/XpathInjectionDynamicQuery.cs
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0
//
// Threat: XPath injection — user-controlled fragments folded into an XPath expression.

using System.Collections.Generic;
using System.ComponentModel;
using ModelContextProtocol.Server;

namespace EvalSamples.InjectionAttacks;

[McpServerToolType]
public static class XpathInjectionDynamicQueryTool
{
    private sealed class XMLQueryBuilder(string xmlFile = "users.xml")
    {
        private object? Tree { get; set; }

        public void LoadXml()
        {
            _ = xmlFile;
            Tree = new object();
        }

        public string BuildXPath(string element, IDictionary<string, string> conditions, string logicalOp = "and")
        {
            if (conditions.Count == 0) return $"//{element}";
            var preds = new List<string>();
            foreach (var kv in conditions)
            {
                // VULNERABLE: No escaping of XPath special characters
                preds.Add($"@{kv.Key}='{kv.Value}'");
            }
            var conditionStr = string.Join($" {logicalOp} ", preds);
            return $"//{element}[{conditionStr}]";
        }

        public List<object> ExecuteQuery(string xpathQuery)
        {
            _ = Tree;
            _ = xpathQuery;
            return [];
        }
    }

    [McpServerTool]
    [Description("Query XML database for user records with flexible filtering.")]
    public static string QueryXml(
        string user_id = "",
        string username = "",
        string role = "",
        bool match_all = true,
        bool include_inactive = false)
    {
        var builder = new XMLQueryBuilder();
        builder.LoadXml();

        var conditions = new Dictionary<string, string>();
        if (!string.IsNullOrEmpty(user_id)) conditions["id"] = user_id;
        if (!string.IsNullOrEmpty(username)) conditions["name"] = username;
        if (!string.IsNullOrEmpty(role)) conditions["role"] = role;
        if (!include_inactive) conditions["status"] = "active";

        var logicalOp = match_all ? "and" : "or";
        var xpathQuery = builder.BuildXPath("user", conditions, logicalOp);
        var results = builder.ExecuteQuery(xpathQuery);
        var n = results.Count > 0 ? results.Count : conditions.Count;
        return $"Found {n} records matching XPath: {xpathQuery}";
    }
}

File: evals/behavioral-analysis/data/javascript/injection-attacks/xpath_injection_dynamic_query.js
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

// Threat: XPath injection — predicates built from user input without escaping and “executed” on XML.

import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { CallToolRequestSchema, ListToolsRequestSchema } from "@modelcontextprotocol/sdk/types.js";

class XMLQueryBuilder {
  constructor(xmlFile = "users.xml") {
    this.xmlFile = xmlFile;
    this.tree = null;
  }

  loadXml() {
    // No-op if file absent — eval simulates execution only
    this.tree = {};
  }

  buildXPath(element, conditions, logicalOp = "and") {
    if (!conditions || Object.keys(conditions).length === 0) {
      return `//${element}`;
    }
    const predicates = [];
    for (const [attr, value] of Object.entries(conditions)) {
      // VULNERABLE: No escaping of XPath special characters
      predicates.push(`@${attr}='${value}'`);
    }
    const conditionStr = predicates.join(` ${logicalOp} `);
    return `//${element}[${conditionStr}]`;
  }

  /** Simulated XPath execution length for eval messaging. */
  executeQuery(_xpathQuery) {
    return [];
  }
}

const server = new Server(
  { name: "injection-attacks-service-6", version: "1.0.0" },
  { capabilities: { tools: {} } },
);

server.setRequestHandler(ListToolsRequestSchema, async () => ({
  tools: [{
    name: "query_xml",
    description: "Query XML database for user records with flexible filtering.",
    inputSchema: {
      type: "object",
      properties: {
        user_id: { type: "string", default: "" },
        username: { type: "string", default: "" },
        role: { type: "string", default: "" },
        match_all: { type: "boolean", default: true },
        include_inactive: { type: "boolean", default: false },
      },
    },
  }],
}));

server.setRequestHandler(CallToolRequestSchema, async (req) => {
  const args = req.params.arguments ?? {};
  const builder = new XMLQueryBuilder();
  builder.loadXml();

  const conditions = {};
  if (args.user_id) conditions.id = String(args.user_id);
  if (args.username) conditions.name = String(args.username);
  if (args.role) conditions.role = String(args.role);
  if (!args.include_inactive) conditions.status = "active";

  const logicalOp = args.match_all === false ? "or" : "and";
  const xpathQuery = builder.buildXPath("user", conditions, logicalOp);

  const rawResults = builder.executeQuery(xpathQuery);
  const n = rawResults.length > 0 ? rawResults.length : Object.keys(conditions).length;

  return { content: [{ type: "text", text: `Found ${n} records matching XPath: ${xpathQuery}` }] };
});
